### DISCUSSION TITLE
local development

### DISCUSSION THREAD

**Discussion by Omegatrixus** [Zaarrg/stremio-community-v5#102] [General]
Hello @Zaarrg, i was wondering, if i wanted to do some development on this , how would i go about it? i am on windows so i suppose i should be using the WebView branch, i have Qt , i just dont know how to spin up a dev environment, will i also need the [stremio-web-shell-fixes](https://github.com/Zaarrg/stremio-web-shell-fixes) ? or the original one? and if so how do i link them up? would i also need the stremio server? i am not really familiar with the new or old stremio system architecture, so i'm trying to figure this out as i go. 

i would really appreciate it if you could direct me to the right way of doing this. thanks in advance!

**Comment by Zaarrg** [comment]
I wouldn't recommend qt6 anymore. WebView2 replaces that build and is now [released](https://github.com/Zaarrg/stremio-desktop-v5/releases). 
And is also way more light weight in deployment and development.

U can check in each branch qt6 (master) and webview2 the `docs` folder which includes a `windows.md` with a guide on how to build.

For the webview2 build this would be the [guide](https://github.com/Zaarrg/stremio-desktop-v5/blob/webview-windows/docs/WINDOWS.md).

Also you dont need [stremio-web-shell-fixes](https://github.com/Zaarrg/stremio-web-shell-fixes) to only spin up the qt6 app or webview2 app. This web ui is used as webui-url instead of the default stremio one as this allows for qt6 specific fixes without potentially breaking the default qt5 shell transport.

And now for WebView2 it includes a WebView2 transport so the webview2 app can work as the offical stremio-web does not have this supported at all as its not open source yet. 
So if u want to see how the app communicates with the web ui then u might need the stremio-web-shell-fixes repo.

**Comment by Omegatrixus** [comment]
Okeeeey, thank you so much for the explanation , i did spin up the WebView2 version, and tried some debugging on the main.cpp to understand a bit how it works,  but i am confused as to what g_webview-> does (as in how to requests and loads a given page or data), and i can't see where the actual ui definitions are, is that what the shell is for? or is it the stremio-runtime.exe that handles this? i didn't check the server.js yet, if so is there a way to have that  in my dev environment too? if its open source that is ahah.

on another note, i see some functions that have no reflection in the program, for instance add extension? how does one trigger that function.

Sorry for bombarding you with such questions.

**Comment by Zaarrg** [comment]
So for starters the app structure would be important to understand. Its basically just layers of windows. Meaning we have MPV as embedded window then on top of that WebView2 which is just the Edge Browser simply said and on top of that the SplashScreen when the app is starting. When a video is played what happens is that the WebView2 window has a Transparent background. And when the web ui in the browser also has a transparent background it shows what's behind it. In this case the mpv embedded window.

And well for stuff like g_webview and what it means you would need to know how webview2 works. This u can do via the webview2 [docs](https://learn.microsoft.com/en-us/microsoft-edge/webview2/webview2-api-reference?tabs=dotnetcsharp).  And this also lists and explains in detail what webview2 related stuff means. For example [here](https://learn.microsoft.com/en-us/microsoft-edge/webview2/samples/webview2apissample#initializewebview).

So in short ``CreateCoreWebView2EnvironmentWithOptions`` together with ``CreateCoreWebView2Controller`` is used  to create the WebView then u can find ``g_webview = coreWebView.try_query<ICoreWebView2_21>();`` this saves the WebView2 interface globally so it can be accessed from everywhere in the app simply said. (It´s not really the interface itself but a smart pointer to the webview2 interface)

For the UI the only thing in the shell is the SplashScreen. As the "real" ui is the [stremio-web-shell-fixes](https://github.com/Zaarrg/stremio-web-shell-fixes) which is a react app. This then has as [WebView Transport](https://github.com/Zaarrg/stremio-web-shell-fixes/blob/development/src/services/Shell/WebViewTransport.js). This allows for the React frontend to communicate with the C++ backend and also listen for incoming messages and can also be called IPC - [Inter-process communication](https://en.wikipedia.org/wiki/Inter-process_communication) as it mimics such behaviour. U can find ``HandleInboundJSON`` which handles inbound communication and ``SendToJS`` which sends to the web ui in the backend ``main.cpp``.

So the actual ui is just a website loaded via ``g_webview->Navigate(g_webuiUrl.c_str());``

Thats why it is also called shell as it doesn't actually have a ui and is more like a skeleton.

The Server.js can be simply downloaded check the ``version/version-details.,json`` this has a link which the updater uses. Though this isn't supposed to be debugged as its a bundled app / server in one single file.

The `stremio-runtime.exe`  is just nodejs packed into a custom exe. As normally u would have something like ``node.exe server.js`` to start the server with node. This way for example when a ``Allow this this app to do network stuff`` popup comes from windows it instead shows that the stremio-runtime.exe is requesting it and not a node.exe. So that's one advantage of that.

Im not sure where u found `add extension` but there is ``SetupExtensions()`` which is called like explained above after the webview is created in the Callback which adds browser extensions if found in the users config folder.

**Comment by Omegatrixus** [comment]
much much appreciated, really, i will be coming back to each of the points you mentioned as i do my own research, that was very helpful, so thank you.

one last thing, i am getting this error, and i wasn't able to find the traces for it or where it originated from  , i suspect its here

https://github.com/Zaarrg/stremio-desktop-v5/blob/1b07eb565fb7bae176bed016f0d1eba8e477c1ac/src/main.cpp#L119-L126

but im not sure, if maybe you have an idea i would be grateful, the app runs without a problem but the assertion error always happens when he app launches.

![Screenshot 2025-01-16 111827](https://github.com/user-attachments/assets/a1b45721-64cb-481f-910a-c633e3e1a2ff)

**Comment by Zaarrg** [comment]
Porbably some file missing or cant be written/read to or from. As it says in the error _osfile & FOPEN would suggest it has something to do with File I/O. 
Make sure when starting the app in ur ``\Debug`` folder that u have stremio-runtime.exe, server.js and libmpv-2.dll. 

Other then that hard to tell u will have to do trial and error. 

Maybe check task manager if stremio, mpv or stremio-runtime is open multiple times. 

The MenuItem code is for the tray menu. If u can right click in tray on stremio and see the menu with no issues then thats not the issue.

Also make sure when running cmake u have used ``-DCMAKE_TOOLCHAIN_FILE=C:\bin\vcpkg\scripts\buildsystems\vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows-static``

And u can also try just deleting the Debug folder and running cmake again.

Also u can try to comment out the autoupdater lines or nodeserver start as those are the only ones that do something with File I/O.

Meaning u can comment out in ``int main(int argc, char* argv[])``
```shell
   # g_updaterThread = std::thread(RunAutoUpdaterOnce);
   # g_updaterThread.detach();
```
And
```shell
 #   if(g_streamingServer){
 #        StartNodeServer();
 #    }

```
